### ISSUE TITLE
linux: keydown/char missing for non-latin keypresses together with control/alt keys

### ISSUE THREAD

**Issue by wxtrac** [wxWidgets/wxWidgets#17643]
*Issue migrated from trac ticket # 17643*

**component:** wxGTK | **priority:** normal

#### 2016-08-29 14:35:50: elfmz created the issue

___
keydown/char events not arrive if input switched to russian if control/alt keys are pressed (Ubuntu, IBus input method).
Also wxAcceleratorTable doesn't catch key combinations (but its less critical for me, just tried as workaround)

However simple test X11 application catches all KeyPress/KeyRelease events in such conditions without problems.

**Comment by unxed** [comment]
This is very interesting issue. With wx 3.1.4, Alt+NonLatinLetter generates KeyUp and KeyDown with empty key code (but it can be deduced by RawKeyFlags which contains hardware keycode under GTK), but no OnChar.

Same is with wx 3.0.4 under Ubuntu 20.04 (GTK 3.24).

But with the same wx 3.0.4 under Ubuntu 18.04 (GTK 3.22) there is no KeyUp/KeyDown at all for such combinations!

@elfmz

**Comment by unxed** [comment]
Okay, 3.0.4 behavior difference between 18.04 and 20.04 is caused by additional backports from 3.0.5 in 20.04, see archive.ubuntu.com/ubuntu/pool/universe/w/wxwidgets3.0/wxwidgets3.0_3.0.4+dfsg-15build1.debian.tar.xz

**Comment by unxed** [comment]
keyup/keydown events absence issue indeed fixed in 3.1 branch.

still I am not sure if it is correct to send 0 as keycode for such combinations. shouldn't we detect keysym by hardware key code instead?

**Comment by vadz** [comment]
> still I am not sure if it is correct to send 0 as keycode for such combinations. shouldn't we detect keysym by hardware key code instead?

The scan code should be already available via `GetRawKeyFlags()`. I'm not sure if we can do anything better, but suggestions are welcome, of course.

Other than that I'm confused: do we still this problem in master or not? [This comment](https://github.com/wxWidgets/wxWidgets/issues/17643#issuecomment-1011151564) seemed to say that we still do, has it been fixed since then?

**Comment by unxed** [comment]
> suggestions are welcome, of course

It seems logical to put Latin key value to keycode field of Alt+NonLatinLetter keydown/up events, as actual local letter resides in UnicodeChar, but sometimes we need to know pressed key, not letter.

To get Latin keycode from GetRawKeyFlags() some translation is needed, like:

```
static int HardwareKeyCodeToKeySym(int code)
{
    // Returns keysym corresponding to passed GTK hardware keycode
    // Returns zero for every key except non-numeric character keys

	switch (code) {

        // Hardware key codes are defined in /usr/share/X11/xkb/keycodes/xfree86

		case 20: return '-';
		case 21: return '=';
		//   22 is Backspace
		//   23 is Tab
		case 24: return 'Q';
		case 25: return 'W';
		case 26: return 'E';
		case 27: return 'R';
		case 28: return 'T';
		case 29: return 'Y';
		case 30: return 'U';
		case 31: return 'I';
		case 32: return 'O';
		case 33: return 'P';
		case 34: return '[';
		case 35: return ']';
		//   36 is Enter
		//   37 is RCtrl
		case 38: return 'A';
		case 39: return 'S';
		case 40: return 'D';
		case 41: return 'F';
		case 42: return 'G';
		case 43: return 'H';
		case 44: return 'J';
		case 45: return 'K';
		case 46: return 'L';
		case 47: return ';';
		case 48: return '\'';
		case 49: return '`';
		//   50 is LShift
		case 51: return '\\';
		case 52: return 'Z';
		case 53: return 'X';
		case 54: return 'C';
		case 55: return 'V';
		case 56: return 'B';
		case 57: return 'N';
		case 58: return 'M';
		case 59: return '.';
		case 60: return '.';
		case 61: return '/';
	}

	return 0; // not applicable
}
```

Also where is strange behavior of `<` key with Shift modifier: its keycode is `<` character in keyup/down events, but it should be `,`. `<` should be in onchar. It is the common behavior with Shift for keys like `>`, `/`, `;`, `'`, etc. But for `<` it's different.

**Comment by unxed** [comment]
> Other than that I'm confused: do we still this problem in master or not?

**Comment by unxed** [comment]
Fixed in #23410

**Comment by unxed** [comment]
> 2016-08-29 14:35:50: elfmz created the issue

@elfmz finally fixed :)
